pkg/dockerregistry/server/token.go

```diff
@@ -0,0 +1,85 @@
+package server
+
+import (
+	"encoding/json"
+	"net/http"
+
+	context "github.com/docker/distribution/context"
+
+	"k8s.io/kubernetes/pkg/client/restclient"
+
+	"github.com/openshift/origin/pkg/client"
+)
+
+type tokenHandler struct {
+	ctx             context.Context
+	anonymousConfig restclient.Config
+}
+
+func NewTokenHandler(ctx context.Context, client *RegistryClient) http.Handler {
+	return &tokenHandler{
+		ctx:             ctx,
+		anonymousConfig: client.SafeClientConfig(),
+	}
+}
+
+// bearer token issued to token requests that present no credentials
+// recognized by the openshift auth provider as identifying the anonymous user
+const anonymousToken = "anonymous"
+
+func (t *tokenHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	ctx := context.WithRequest(t.ctx, req)
+
+	// If no authorization is provided, return a token the auth provider will treat as an anonymous user
+	if len(req.Header.Get("Authorization")) == 0 {
+		context.GetRequestLogger(ctx).Debugf("anonymous token request")
+		t.writeToken(anonymousToken, w, req)
+		return
+	}
+
+	// use the password as the token
+	_, token, ok := req.BasicAuth()
+	if !ok {
+		context.GetRequestLogger(ctx).Debugf("no basic auth credentials provided")
+		t.writeUnauthorized(w, req)
+		return
+	}
+
+	// TODO: if this doesn't validate as an API token, attempt to obtain an API token using it?
+	copied := t.anonymousConfig
```

Comment:
In case you see it elsewhere, remember that this doesn't work in the general case.  `Config` contains `AuthProvider` which is a pointer and contains a `map`.  The anonymous case should have those fields stripped, so you're good here.